engine/crates/engine-v2/id-derives/src/id.rs

```diff
@@ -27,11 +27,27 @@ pub fn derive_id(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
     ));
 
     let output = quote! {
+        impl #impl_generics From<i32> for #ident #ty_generics #where_clause {
+            fn from(value: i32) -> Self {
+                Self::from(
+                    #inner_ty::try_from(value).expect("Value mismatch")
+                )
+            }
+        }
+
         impl #impl_generics From<usize> for #ident #ty_generics #where_clause {
             fn from(value: usize) -> Self {
+                Self::from(
+                    #inner_ty::try_from(value).expect("Too big")
+                )
+            }
+        }
+
+        impl #impl_generics From<#inner_ty> for #ident #ty_generics #where_clause {
+            fn from(value: #inner_ty) -> Self {
                 #max_check
                 Self(
-                    #inner_ty::try_from(value + 1).ok().and_then(|value| value.try_into().ok()).expect(#too_many_error)
+                    (value + 1).try_into().expect(#too_many_error)
                 )
             }
```

Comment:
I may be missing something, but why are we only doing the `+ 1` in the `From<#inner_ty>` impl?  That kind of inconsistency seems like it'd cause off by one errors...

Also why do we need a conversion from i32 _at all_ - surely we never need negative numbers?

Reply:
I might have done stupid stuff somewhere, but somehow it didn't accept `ObjectDefinition::from(0)`, got annoyed and added it. But after the coercion fix I'm not creating as many of those anyway. So removing it, you're right.

The `+1` was always done in the `From` until now, are you suggesting creating a `new` method? Or do I miss something?

Reply:
> The +1 was always done in the From until now, are you suggesting creating a new method? Or do I miss something?

Of the three impls I have highlighted only the last one contains a `+ 1`.  The first two do not - that seems wrong?

Reply:
the first two use the last one. so `From<usize>` does: `Self::from(u32::try_from(value))`

Reply:
Hmm...  Could be clearer tbh.

Reply:
yeah agreed, it's not super obvious in the macro